src/Libraries/Microsoft.Extensions.AI.Evaluation.NLP/Common/BLEUAlgorithm.cs

```diff
@@ -0,0 +1,197 @@
+        foreach (var h in hypCounts)
+        {
+            if (maxCounts.TryGetValue(h.Key, out var v))
+            {
+                clippedCounts[h.Key] = Math.Min(h.Value, v);
+            }
+            else
+            {
+                // If the hypothesis n-gram is not in any reference, it is clipped to 0.
+                clippedCounts[h.Key] = 0;
+            }
+        }
+
+        int numerator = clippedCounts.Values.Sum();
+        int denominator = Math.Max(1, hypCounts.Sum());
+
+        return new RationalNumber(numerator, denominator);
+    }
+
+    /// <summary>
+    /// Generate an n-sized array of equal weights that sum to 1.0.
+    /// </summary>
+    /// <param name="n">Number of weights to return.</param>
+    /// <returns>Array of equal sized values that sum to 1.0.</returns>
+    internal static double[] EqualWeights(int n)
+    {
+        if (n <= 0)
+        {
+            Throw.ArgumentOutOfRangeException(nameof(n), $"'{nameof(n)}' must be greater than zero.");
+        }
+
+        double[] weights = new double[n];
+        for (int i = 0; i < n; i++)
+        {
+            weights[i] = 1.0 / n;
+        }
+
+        return weights;
+    }
+
+    internal static readonly double[] DefaultBLEUWeights = EqualWeights(4);
+
+    internal static double SentenceBLEU(IEnumerable<IEnumerable<string>> references, IEnumerable<string> hypothesis,
+        double[]? weights = null, Func<RationalNumber[], int, double[]>? smoothingFunction = null)
+    {
+        if (references == null || !references.Any())
+        {
+            Throw.ArgumentNullException(nameof(references), $"'{nameof(references)}' cannot be null or empty.");
+        }
+
+        if (hypothesis == null || !hypothesis.Any())
+        {
+            Throw.ArgumentNullException(nameof(hypothesis), $"'{nameof(hypothesis)}' cannot be null or empty.");
+        }
+
+        if (weights is null)
+        {
+            weights = DefaultBLEUWeights;
+        }
+
+        if (weights.Length == 0)
+        {
+            Throw.ArgumentNullException(nameof(weights), $"'{nameof(weights)}' cannot be empty.");
+        }
+
+        var precisionValues = new RationalNumber[weights.Length];
+        for (int i = 0; i < weights.Length; i++)
+        {
+            int n = i + 1;
+            RationalNumber prec = ModifiedPrecision(references, hypothesis, n);
+
+            if (i == 0 && prec.Numerator == 0)
+            {
+                // If the precision for unigrams (n == 1) is zero, the there can be no higher order matches and BLEU score is zero.
+                return 0.0;
+            }
+
+            precisionValues[i] = prec;
+        }
+
+        int hypLen = hypothesis.Count();
+        int closestRefLength = ClosestRefLength(references, hypLen);
+        double brevityPenalty = BrevityPenalty(closestRefLength, hypLen);
+
+        if (smoothingFunction == null)
+        {
+            smoothingFunction = SmoothingFunction.Method0;
```

Comment:
Looks like the evaluator always passes `SmoothingFunction.Method4`. Should the default be `Method4` rather than `Method0` (and then we can also omit passing the smoothing function in the evaluator)?

Reply:
Conceptually, Method0 is no smoothing. I want to remain true to the original implementation and reflect the fact that there should be no smoothing by default, but that we choose to use Method4 as our default.